### PR TITLE
Add header_cache_control to query stats

### PR DESCRIPTION
**Pull request by krajorama**
#### What this PR does

Mimir continuous test executes queries with and without results cache enabled via the Cache-Control header. It's hard to tell in the query-stats which is which since everything else is the same.

Add `header_cache_control=...` to query stats.

#### Which issue(s) this PR fixes or relates to

Fixes N/A

#### Checklist

- [x] Tests updated.
- N/A Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- N/A [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/frontend/transport/handler.go [hunk 1]
```diff
@@ -347,6 +347,9 @@ func (f *Handler) reportQueryStats(
 			"results_cache_hit_bytes", details.ResultsCacheHitBytes,
 			"results_cache_miss_bytes", details.ResultsCacheMissBytes,
 		)
+		if details.ResultsCacheNoStore {
+			logMessage = append(logMessage, "results_cache_no_store", true)
```
Comment (pracucci) [inline pkg/frontend/transport/handler.go:351]: I suggest to always add the log field, not just when it's true. I think it makes logs easier to parse.

**Review by pracucci — DISMISSED** [review]

File: pkg/frontend/transport/handler.go (hunk 1 above)
Comment (krajorama) [reply pkg/frontend/transport/handler.go]: Tested locally :

header not present:

`curl -G   http://localhost:8001/prometheus/api/v1/query_range --data-urlencode 'query=up{x="y"}' --data-urlencode 'start=1719993614' --data-urlencode 'end=1719995414' --data-urlencode 'step=1m'`

```
mimir-1-1        | ts=2024-07-03T08:32:10.867257807Z caller=handler.go:383 level=info user=anonymous traceID=6ee0dc648a604a48 msg="query stats" component=query-frontend method=GET path=/prometheus/api/v1/query_range route_name=prometheus_api_v1_query_range user_agent=curl/7.68.0 status_code=200 response_time=7.371747ms response_size_bytes=63 query_wall_time_seconds=0.005392547 fetched_series_count=0 fetched_chunk_bytes=0 fetched_chunks_count=0 fetched_index_bytes=0 sharded_queries=0 split_queries=1 estimated_series_count=0 queue_time_seconds=5.35e-05 param_step=60000 param_query="up{x=\"y\"}" param_start=2024-07-03T08:00:14Z param_end=2024-07-03T08:30:14Z length=35m0s time_since_min_time=36m56.859773194s time_since_max_time=1m56.859773194s results_cache_hit_bytes=0 results_cache_miss_bytes=81 
results_cache_no_store=false status=success
```

header present
`curl -G -H "Cache-Control: no-store"   http://localhost:8001/prometheus/api/v1/query_range --data-urlencode 'query=up{x="y"}' --data-urlencode 'start=1719993614' --data-urlencode 'end=1719995414' --data-urlencode 'step=1m'`


```
mimir-1-1        | ts=2024-07-03T08:32:43.265653417Z caller=handler.go:383 level=info user=anonymous traceID=51702283e2e53281 msg="query stats" component=query-frontend method=GET path=/prometheus/api/v1/query_range route_name=prometheus_api_v1_query_range user_agent=curl/7.68.0 status_code=200 response_time=5.211026ms response_size_bytes=63 query_wall_time_seconds=0.003876335 fetched_series_count=0 fetched_chunk_bytes=0 fetched_chunks_count=0 fetched_index_bytes=0 sharded_queries=0 split_queries=1 estimated_series_count=0 queue_time_seconds=5.0635e-05 param_end=2024-07-03T08:30:14Z param_step=60000 param_query="up{x=\"y\"}" param_start=2024-07-03T08:00:14Z length=35m0s time_since_min_time=37m29.260357532s time_since_max_time=2m29.260357532s results_cache_hit_bytes=0 results_cache_miss_bytes=81 
results_cache_no_store=true status=success
```

**Comment by pracucci** [comment]
> there are the fields which show the number of bytes retried before storing in the cache - results_cache_hit_bytes and results_cache_miss_bytes. Doesn't this work for you?

It doesn't tell if looking up the cache was skipped or it was done but it was a cache miss.

**Comment by dimitarvdimitrov** [comment]
> > there are the fields which show the number of bytes retried before storing in the cache - results_cache_hit_bytes and results_cache_miss_bytes. Doesn't this work for you?
> 
> It doesn't tell if looking up the cache was skipped or it was done but it was a cache miss.

i'm missing some background for this. What are you trying to find out with this log field (any why?)

**Comment by krajorama** [comment]
> i'm missing some background for this. What are you trying to find out with this log field (any why?)

The issue is that continuous test runs the same query with and without cache enabled (header Cache-Control: no-store). When there's no error, the continuous test doesn't print anything so on the query frontend it's kind of hard to know whether we got  results_cache_hit_bytes=0 due to an error or if the cache was turned off.

**Comment by dimitarvdimitrov** [comment]
aha, I see how including this header will make debugging easier. What's your take on the `header_` prefix?

**Comment by krajorama** [comment]
> aha, I see how including this header will make debugging easier. What's your take on the `header_` prefix?

With cache control set:
```
mimir-1-1        | ts=2024-07-08T15:36:17.3673853Z caller=handler.go:384 level=info user=anonymous traceID=76ee71b5e41f2e1c msg="query stats" component=query-frontend method=GET path=/prometheus/api/v1/query_range route_name=prometheus_api_v1_query_range user_agent=curl/7.68.0 status_code=200 response_time=20.526178ms response_size_bytes=63 query_wall_time_seconds=0.01887049 fetched_series_count=0 fetched_chunk_bytes=0 fetched_chunks_count=0 fetched_index_bytes=0 sharded_queries=0 split_queries=1 estimated_series_count=0 queue_time_seconds=3.8064e-05 
header_cache_control=no-store param_end=2024-07-03T08:30:14Z param_step=60000 param_query="up{x=\"y\"}"
param_start=2024-07-03T08:00:14Z length=35m0s time_since_min_time=127h41m3.346759108s time_since_max_time=127h6m3.346759108s results_cache_hit_bytes=0 results_cache_miss_bytes=81 status=success
```

Not set:
```
mimir-1-1        | ts=2024-07-08T15:37:55.184389345Z caller=handler.go:384 level=info user=anonymous traceID=2a8f918113a97abe msg="query stats" component=query-frontend method=GET path=/prometheus/api/v1/query_range route_name=prometheus_api_v1_query_range user_agent=curl/7.68.0 status_code=200 response_time=4.389467ms response_size_bytes=63 query_wall_time_seconds=0.002727982 fetched_series_count=0 fetched_chunk_bytes=0 fetched_chunks_count=0 fetched_index_bytes=0 sharded_queries=0 split_queries=1 estimated_series_count=0 queue_time_seconds=6.0414e-05 
header_cache_control= param_step=60000 param_query="up{x=\"y\"}" param_start=2024-07-03T08:00:14Z
param_end=2024-07-03T08:30:14Z length=35m0s time_since_min_time=127h42m41.179900912s time_since_max_time=127h7m41.179900912s results_cache_hit_bytes=0 results_cache_miss_bytes=81 status=success
```

File: pkg/frontend/transport/handler.go [hunk 2]
```diff
@@ -264,6 +264,7 @@ func (f *Handler) reportSlowQuery(r *http.Request, queryString url.Values, query
 		"host", r.Host,
 		"path", r.URL.Path,
 		"time_taken", queryResponseTime.String(),
+		"header_cache_control", r.Header.Get("Cache-Control"),
```
Comment (dimitarvdimitrov) [inline pkg/frontend/transport/handler.go:267]: tiny nitpick: it's easier to discover in the code if it's added in the same place other headers are logged. It also won't be duplicated:
https://github.com/grafana/mimir/blob/9ef67470f516f6b2567e5d44e4f21772276a99da/pkg/frontend/transport/handler.go#L422-L429

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM

File: pkg/frontend/transport/handler.go (hunk 2 above)
Comment (krajorama) [reply pkg/frontend/transport/handler.go]: Had to refactor  a little. On the bright side we now have unit tests that check that `-query-frontend.log-query-request-headers` actually works.

**Review by dimitarvdimitrov — APPROVED** [review]
thanks for all the back and forth 🙏

**Pull request merged**